File: fuzz/python/json_parser_fuzzer.py
#!/usr/bin/env python3
"""
JSON Parser Fuzzer using Atheris
Tests JSON parsing robustness with malformed input data
"""

import atheris
import json
import sys
from typing import Any, Dict, List, Union


def test_json_parsing(data: bytes) -> None:
    """Test JSON parsing with various malformed inputs"""
    
    # Test with UTF-8 string conversion
    try:
        json_str = data.decode('utf-8')
    except UnicodeDecodeError:
        # Skip invalid UTF-8 data
        return
    
    # Test basic JSON parsing
    try:
        parsed = json.loads(json_str)
        # Test JSON serialization back to string
        _ = json.dumps(parsed)
    except (json.JSONDecodeError, TypeError, ValueError, OverflowError):
        # Expected errors for malformed JSON
        pass
    
    # Test with truncated data
    if len(data) > 1:
        truncated = data[:-1]
        try:
            truncated_str = truncated.decode('utf-8')
            _ = json.loads(truncated_str)
        except (UnicodeDecodeError, json.JSONDecodeError):
            pass
    
    # Test with corrupted data
    if len(data) > 10:
        corrupted = bytearray(data)
        # Flip some bits
        for i in range(min(10, len(corrupted)):
            corrupted[i] = (corrupted[i] + 1) % 256
        
        try:
            corrupted_str = corrupted.decode('utf-8')
            _ = json.loads(corrupted_str)
        except (UnicodeDecodeError, json.JSONDecodeError):
            pass
    
    # Test with nested structures
    if len(data) > 20:
        try:
            # Create complex JSON structure
            complex_json = f'''{{
                "string": "{json_str[:10]}",
                "number": {len(data)},
                "boolean": {len(data) % 2 == 0},
                "null": null,
                "array": [1, 2, 3],
                "object": {{"nested": "value"}}
            }}'''
            _ = json.loads(complex_json)
        except (json.JSONDecodeError, TypeError):
            pass
    
    # Test with very large numbers
    if len(data) > 0:
        try:
            large_number = str(len(data) * 1000000)
            _ = json.loads(large_number)
        except (json.JSONDecodeError, ValueError):
            pass
    
    # Test with escaped strings
    if len(data) > 0:
        try:
            escaped_string = f'"{json_str.replace("\\", "\\\\").replace('"', '\\"')}"'
            _ = json.loads(escaped_string)
        except (json.JSONDecodeError, ValueError):
            pass
    
    # Test with unicode data
    if len(data) > 0:
        try:
            unicode_string = f'"{json_str.encode("unicode_escape").decode()}"'
            _ = json.loads(unicode_string)
        except (json.JSONDecodeError, ValueError):
            pass


def test_json_parsing_edge_cases(data: bytes) -> None:
    """Test JSON parsing with edge cases"""
    
    # Test with empty data
    if len(data) == 0:
        try:
            _ = json.loads("")
        except json.JSONDecodeError:
            pass
    
    # Test with single character
    if len(data) == 1:
        try:
            single_char = data.decode('utf-8', errors='ignore')
            _ = json.loads(single_char)
        except (UnicodeDecodeError, json.JSONDecodeError):
            pass
    
    # Test with very large data
    if len(data) > 1024:
        try:
            large_str = data.decode('utf-8', errors='ignore')
            _ = json.loads(large_str)
        except (UnicodeDecodeError, json.JSONDecodeError):
            pass


def test_json_parsing_formats(data: bytes) -> None:
    """Test JSON parsing with different formats"""
    
    try:
        json_str = data.decode('utf-8')
    except UnicodeDecodeError:
        return
    
    # Test with different JSON formats
    formats = [
        json_str,
        json_str.replace('"', "'"),  # Single quotes
        json_str.replace('true', 'True'),  # Python boolean
        json_str.replace('false', 'False'),  # Python boolean
        json_str.replace('null', 'None'),  # Python None
    ]
    
    for fmt in formats:
        try:
            _ = json.loads(fmt)
        except (json.JSONDecodeError, ValueError):
            pass


def main():
    """Main fuzzing function"""
    
    def fuzz_json(data: bytes) -> None:
        """Fuzz JSON parsing functions"""
        test_json_parsing(data)
        test_json_parsing_edge_cases(data)
        test_json_parsing_formats(data)
    
    # Set up Atheris fuzzer
    atheris.Setup(sys.argv, fuzz_json)
    atheris.Fuzz()


if __name__ == "__main__":
    main()

File: fuzz/rust/src/json_parser_fuzzer.rs
#![no_main]

use libfuzzer_sys::fuzz_target;
use serde_json::{Value, from_str, from_slice, from_reader};
use std::io::Cursor;

fuzz_target!(|data: &[u8]| {
    // Test JSON string parsing
    if let Ok(json_str) = std::str::from_utf8(data) {
        let _ = from_str::<Value>(json_str);
    }

    // Test JSON slice parsing
    let _ = from_slice::<Value>(data);

    // Test JSON reader parsing
    let cursor = Cursor::new(data);
    let _ = from_reader::<Value, Cursor<&[u8]>>(cursor);

    // Test with truncated data
    if data.len() > 1 {
        let truncated = &data[..data.len() - 1];
        if let Ok(json_str) = std::str::from_utf8(truncated) {
            let _ = from_str::<Value>(json_str);
        }
        let _ = from_slice::<Value>(truncated);
    }

    // Test with corrupted data
    if data.len() > 10 {
        let mut corrupted = data.to_vec();
        // Flip some bits
        for i in 0..std::cmp::min(10, data.len()) {
            corrupted[i] = corrupted[i].wrapping_add(1);
        }
        
        if let Ok(json_str) = std::str::from_utf8(&corrupted) {
            let _ = from_str::<Value>(json_str);
        }
        let _ = from_slice::<Value>(&corrupted);
    }

    // Test with nested structures
    if data.len() > 20 {
        // Create a complex JSON structure
        let complex_json = format!(
            r#"{{
                "string": "{}",
                "number": {},
                "boolean": {},
                "null": null,
                "array": [1, 2, 3],
                "object": {{"nested": "value"}}
            }}"#,
            std::str::from_utf8(data).unwrap_or("corrupted"),
            data.len(),
            data.len() % 2 == 0
        );
        
        let _ = from_str::<Value>(&complex_json);
    }

    // Test with very large numbers
    if data.len() > 0 {
        let large_number = format!("{}", data.len() * 1000000);
        let _ = from_str::<Value>(&large_number);
    }

    // Test with escaped strings
    if data.len() > 0 {
        let escaped_string = format!(
            r#""{}""#,
            std::str::from_utf8(data)
                .unwrap_or("corrupted")
                .replace("\\", "\\\\")
                .replace("\"", "\\\"")
        );
        let _ = from_str::<Value>(&escaped_string);
    }

    // Test with unicode data
    if data.len() > 0 {
        let unicode_string = format!(
            r#""{}""#,
            std::str::from_utf8(data)
                .unwrap_or("corrupted")
                .chars()
                .map(|c| format!("\\u{:04x}", c as u32))
                .collect::<String>()
        );
        let _ = from_str::<Value>(&unicode_string);
    }
});
